docs/architecture/adr-065-custom-event-indexing.md

```diff
@@ -0,0 +1,341 @@
+indexed by the respective type.
+
+In addition to indexing, Tendermint Core also supports the ability to query for
+both indexed transaction and block events via Tendermint's RPC layer. The ability
+to query for these indexed events facilitates a great multitude of upstream client
+and application capabilities, e.g. block explorers, IBC relayers, and auxiliary
+data availability and indexing services.
+
+Currently, Tendermint only supports indexing via a `kv` indexer, which is supported
+by an underlying embedded key/value store database. The `kv` indexer implements
+its own indexing and query mechanisms. While the former is somewhat trivial,
+providing a rich and flexible query layer is not as trivial and has caused many
+issues and UX concerns for upstream clients and applications.
+
+The fragile nature of the proprietary `kv` query engine and the potential
+performance and scaling issues that arise when a large number of consumers are
+introduced, motivate the need for a more robust and flexible indexing and query
+solution.
+
+## Alternative Approaches
+
+With regards to alternative approaches to a more robust solution, the only serious
+contender that was considered was to transition to using [SQLite](https://www.sqlite.org/index.html).
+
+While the approach would work, it locks us into a specific query language and
+storage layer, so in some ways it's only a bit better than our current approach.
+In addition, the implementation would require the introduction of CGO into the
+Tendermint Core stack, whereas right now CGO is only introduced depending on
+the database used.
+
+## Decision
+
+We will adopt a similar approach to that of the Cosmos SDK's `KVStore` state
+listening described in [ADR-038](https://github.com/cosmos/cosmos-sdk/blob/master/docs/architecture/adr-038-state-listening.md).
+
+Namely, we will perform the following:
+
+- Introduce a new interface, `EventSink`, that all data sinks must implement.
+- Augment the existing `tx_index.indexer` configuration to now accept a series
+  of one or more indexer types, i.e sinks.
+- Combine the current `TxIndexer` and `BlockIndexer` into a single `KVEventSink`
+  that implements the `EventSink` interface.
+- Introduce an additional `EventSink` that is backed by [PostgreSQL](https://www.postgresql.org/).
+  - Implement the necessary schemas to support both block and transaction event
+  indexing.
+- Update `IndexerService` to use a series of `EventSinks`.
+- Proxy queries to the relevant sink's native query layer.
+- Update all relevant RPC methods.
+
+
+## Detailed Design
+
+### EventSink
+
+We introduce the `EventSink` interface type that all supported sinks must implement.
+The interface is defined as follows:
+
+```go
+type EventSink interface {
+  IndexBlockEvents(types.EventDataNewBlockHeader) error
+  IndexTxEvents(*abci.TxResult) error
+
+  SearchBlockEvents(context.Context, *query.Query) ([]int64, error)
+  SearchTxEvents(context.Context, *query.Query) ([]*abci.TxResult, error)
```

Comment:
It feels wrong to me to pull this search interface directly into the application/node process.  The search/query interface of this data seems like it should be handled by an external system or at least a separate process.  The indexing and querying semantics are a completely separate concern.  As a good example if a Kafka system or even a time series database were wired up to sink these events the query format would be massively different than if Postgres were used.  The smell of "use a string for this query" above should be a strong clue that very little universal constraint can be applied here and remain useful.

Reply:
Generally agree with this. I found it confusing especially when a single node can have multiple event sinks. Maybe tx_search is only supported with KV event sink, everything else you use the external resource for searching?

Reply:
I also agree with Ira.

It would be good to support the "legacy" mode for a smooth transition, but it would be just like supporting amino signing mode - an interim solution. In fact, I think the whole "Tendermint handles indexing and querying for you" has been proven to be a failure. Tendermint refused to add any external dbs in order to keep the binaries lean and without external dependencies, which is a fair choice as a way to build a blockchain (not an indexing node). But it also means it will never provide a good indexing service.

I think we need to make a decision there. I support providing such an indexing service (to provide support for external indexers) and then let the community create a few different open source variants. Ideally these should standardize on 2 or 3 APIs for different domains (eg. GraphQL, Time Series, PostgreSQL tables). But the standarization must come after experimentation - let the best indexer win.

Since you can plug any indexer into any Cosmos SDK blockchain, it just requires running your own full node to add such support.

Reply:
Agree with Ira and Ethan,
Even for the tx search, I think it is better to handle it by the external system. The core should just provide a simple tx or event query. I feel like the current tx_search implementation can be an attack factor to the core. However, supporting the "legacy" is also important in the current status. but might need to put more constraint (and document it) on the query conditions.

Reply:
Yes, searching should obviously be offloaded to the sink, however, we keep the search interface(s) due to the need for the legacy `KV` indexer.

People will lose their minds if all of a sudden they can't search for blocks or txs by events. 95% of operators will have absolutely zero interest in spinning up and managing a postgres cluster, so hence the motivation for keeping the `KV` indexer which requires this interface.

That being said, all other indexers, `psql` for the time being, can simply just error on search.